### PR TITLE
Adding cloud provider OpenShift CNV (OpenShift Virtualization)

### PR DESCRIPTION
**Pull request by agonzalezrh**
Please double check everything

### CODE REVIEW

File: ansible/cloud_providers/openshift_cnv_destroy_env.yml
Comment (rut31337) [inline ansible/cloud_providers/openshift_cnv_destroy_env.yml:1]: shouldnt this be in the execution environment?

File: ansible/cloud_providers/openshift_cnv_destroy_env.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+- import_playbook: ../include_vars.yml
+
+- name: Delete Infrastructure
+  hosts: localhost
+  connection: local
+  gather_facts: false
+  become: false
+  tasks:
+    - name: Install community.okd collection for openshift modules
+      command: >-
```
Comment (wkulhanek) [inline ansible/cloud_providers/openshift_cnv_destroy_env.yml]: This should be in the EE (it is in :latest for me).

File: ansible/cloud_providers/openshift_cnv_infrastructure_deployment.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    - infra
+    - infra_tasks
+  module_defaults:
+    group/k8s:
+      host: "{{ openshift_cnv_api_uri }}"
+      validate_certs: false
+
+  tasks:
+    - name: Install community.okd collection for openshift modules
+      command: >-
```
Comment (wkulhanek) [inline ansible/cloud_providers/openshift_cnv_infrastructure_deployment.yml]: Same here?

File: ansible/cloud_providers/openshift_cnv_infrastructure_deployment.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        ansible-galaxy collection install
+        -p collections community.okd
+        --force-with-deps
+      args:
+        chdir: "{{ playbook_dir }}/.."
+        creates: "{{ playbook_dir }}/../collections/ansible_collections/community/okd"
+
+    - name: Create ssh provision key
+      include_role:
```
Comment (wkulhanek) [inline ansible/cloud_providers/openshift_cnv_infrastructure_deployment.yml]: ansible.builtin.include_role

File: ansible/roles-infra/infra-openshift-cnv-resources/defaults/main.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+---
+openshift_cnv_project_name: "{{ env_type }}-{{ guid }}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/defaults/main.yaml:2]: `uuid`?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/check_network.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+---
+- name: Get Network if exists
+  kubernetes.core.k8s_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/check_network.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+          name: "INSTANCENAME-{{guid}}"
+        name: "INSTANCENAME-{{guid}}"
+    _instance_disks:
+      - disk:
+          bus: "{{ _instance.disk_type | default('virtio') }}"
+        name: "INSTANCENAME-{{guid}}"
+
+- name: Set the instances interfaces
+  ansible.builtin.set_fact:
+    _instance_interfaces: "{{ _instance_interfaces + [{'name': _network, 'macAddress': _instance.fixed_macs[_network]|default('2c:c2:60'|random_mac),'bridge': {}, 'model': 'e1000e', 'pciAddress': '0000:00:0' + (_network_idx+3)|string + '.0'} if _network != 'default' else {'name': 'default', 'model': 'e1000e', 'pciAddress': '0000:00:0' + (_network_idx+3)|string + '.0', 'masquerade': {}}] }}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml:17]: formatting difficult to read

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    _instance_networks: "{{ _instance_networks + [{'name': _network, 'multus': {'networkName': _network + guid}} if _network != 'default' else {'name': 'default', 'pod': {}}] }}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml:18]: formatting

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  loop: "{{ _instance.networks | default([]) | list }}"
+  loop_control:
+    loop_var: _network
+    index_var: _network_idx
+
+- name: Set the instances disks
+  ansible.builtin.set_fact:
+    _instance_disks: "{{ _instance_disks|from_yaml + [{'name': _disk.metadata.name, 'disk': {'bus': _instance.disk_type|default('virtio')} }] }}"
+    _instance_volumes: "{{ _instance_volumes|from_yaml + [{'name': _disk.metadata.name, 'dataVolume': {'name': _disk.metadata.name} }] }}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml:27]: formatting

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    _cloud_config: |-
+        #cloud-config
+        ssh_authorized_keys:
+        - {{ lookup('file', openshift_cnv_ssh_pub_key_path ) }}
+        {{_instance.userdata|default('')|replace('#cloud-config','')|default('')}}
+
+- name: Set cloud init disk if needed
+  set_fact:
+    _instance_volumes: "{{ _instance_volumes|from_yaml + [{ 'name': 'cloudinitdisk', 'cloudInitNoCloud': {'userDataBase64': _cloud_config|b64encode,
+     'networkData': _instance.networkdata|default('')}}] }}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_instance.yaml:44]: formatting

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_network.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+---
+- name: Create network {{ _network.name }}
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_network.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_network.yaml [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+    definition:
+      apiVersion: v1
+      kind: NetworkAttachmentDefinition
+      metadata:
+        name: "{{ _network.name }}{{ guid }}"
+        namespace: "{{ openshift_cnv_project_name }}"
+      spec:
+        config: "{{ config | to_json }}"
+  vars:
+    config: "{'cniVersion':'0.3.1','type':'ovn-k8s-cni-overlay','topology':'layer2','name': '{{ _network.name }}{{ guid }}', 'netAttachDefName': '{{ openshift_cnv_project_name }}/{{ _network.name }}{{ guid }}', 'mtu': 9000}"
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_network.yaml:13]: formatting

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_project.yaml [hunk 12]
```diff
@@ -0,0 +1,39 @@
+---
+- name: Create OpenShift CNV project
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_project.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_project.yaml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    definition:
+      metadata:
+        labels:
+          mutatepods.kubemacpool.io: ignore
+          mutatevirtualmachines.kubemacpool.io: ignore
+  register: r_project
+
+
+- name: Allow clone PVC from cnv-images namespace to the project namespace
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_project.yaml:17]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_routes.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
+---
+- name: Create routes with tls
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_routes.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_routes.yaml [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        to:
+          kind: Service
+          name: "{{ route.service }}"
+  when: "{{ route.tls|default(False) == True}}"
+  loop: "{{ _instance.routes|default([]) }}"
+  loop_control:
+    loop_var: route
+
+- name: Create routes without tls
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_routes.yaml:26]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_service.yaml [hunk 16]
```diff
@@ -0,0 +1,20 @@
+---
+- name: Create services for instance "{{ _instance.name }}"
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_service.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_services.yaml [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        namespace: "{{ openshift_cnv_project_name }}"
+      spec:
+        ports:
+        - port: {{ _instance.servicePort | default(22) |int}}
+          protocol: TCP
+          targetPort: {{ _instance.servicePort | default(22)|int }}
+        selector:
+          vm.cnv.io/name: "{{ _instance_name }}"
+
+  kubernetes.core.k8s:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/create_services.yaml:19]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/delete_networks.yaml [hunk 18]
```diff
@@ -0,0 +1,6 @@
+---
+- name: Get a list of Network
+  kubernetes.core.k8s_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/delete_networks.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/start_instances.yaml [hunk 19]
```diff
@@ -0,0 +1,18 @@
+---
+- name: Get a list of VMs
+  kubernetes.core.k8s_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/start_instances.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/status_instances.yaml [hunk 20]
```diff
@@ -0,0 +1,25 @@
+---
+- name: Get a list of VMs
+  kubernetes.core.k8s_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/status_instances.yaml:3]: retries?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/status_instances.yaml [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+    kind: VirtualMachine
+    namespace: "{{ openshift_cnv_project_name }}"
+  register: r_vm_list
+
+- name: Report status in user info
+  agnosticd_user_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/status_instances.yaml:9]: what about `agnosticd_user_data`?

File: ansible/roles-infra/infra-openshift-cnv-resources/tasks/stop_instances.yaml [hunk 22]
```diff
@@ -0,0 +1,19 @@
+---
+- name: Get a list of VMIs
+  kubernetes.core.k8s_info:
```
Comment (fridim) [inline ansible/roles-infra/infra-openshift-cnv-resources/tasks/stop_instances.yaml:3]: retries?

**Review by fridim — CHANGES_REQUESTED** [review]
Thanks for the PR.

I'm wondering if there is a way to specify retries/wait at the k8s module level for all those call to the openshift API, or if we have to deal with it at the ansible level with the classical `register + retries + until`

Update the documentation in `docs/` to add the new cloud-provider?

**Review by fridim — APPROVED** [review]

**Pull request merged**